viam-cartographer.go

```diff
@@ -124,92 +194,311 @@ func New(
 	if cartoSvc.primarySensorName != "" {
-		if err := dim2d.ValidateGetAndSaveData(cancelCtx, cartoSvc.dataDirectory, lidar,
-			sensorValidationMaxTimeoutSec, sensorValidationIntervalSec, cartoSvc.logger); err != nil {
-			return nil, errors.Wrap(err, "getting and saving data failed")
+		if err := dim2d.ValidateGetAndSaveData(cancelCtx, cartoSvc.dataDirectory, cartoSvc.lidar,
+			cartoSvc.sensorValidationMaxTimeoutSec, cartoSvc.sensorValidationIntervalSec, cartoSvc.logger); err != nil {
+			return errors.Wrap(err, "getting and saving data failed")
 		}
-		cartoSvc.StartDataProcess(cancelCtx, lidar, nil)
-		logger.Debugf("Reading data from sensor: %v", cartoSvc.primarySensorName)
+		cartoSvc.StartDataProcess(cancelCtx, cartoSvc.lidar, nil)
+		cartoSvc.logger.Debugf("Reading data from sensor: %v", cartoSvc.primarySensorName)
 	}
 
 	if err := cartoSvc.StartSLAMProcess(ctx); err != nil {
-		return nil, errors.Wrap(err, "error with slam service slam process")
+		return errors.Wrap(err, "error with slam service slam process")
 	}
 
-	client, clientClose, err := vcConfig.SetupGRPCConnection(ctx, cartoSvc.port, dialMaxTimeoutSec, logger)
+	client, clientClose, err := vcConfig.SetupGRPCConnection(ctx, cartoSvc.port, cartoSvc.dialMaxTimeoutSec, cartoSvc.logger)
 	if err != nil {
-		return nil, errors.Wrap(err, "error with initial grpc client to slam algorithm")
+		return errors.Wrap(err, "error with initial grpc client to slam algorithm")
 	}
 	cartoSvc.clientAlgo = client
 	cartoSvc.clientAlgoClose = clientClose
 
-	success = true
-	return cartoSvc, nil
+	return nil
 }
 
 // cartographerService is the structure of the slam service.
 type cartographerService struct {
 	resource.Named
 	resource.AlwaysRebuild
 	primarySensorName string
+	lidar             lidar.Lidar
 	executableName    string
 	subAlgo           SubAlgo
-	slamProcess       pexec.ProcessManager
-	clientAlgo        pb.SLAMServiceClient
-	clientAlgoClose   func() error
-
-	configParams        map[string]string
-	dataDirectory       string
+	// deprecated
+	slamProcess pexec.ProcessManager
+	// deprecated
+	clientAlgo pb.SLAMServiceClient
+	// deprecated
+	clientAlgoClose func() error
+
+	configParams  map[string]string
+	dataDirectory string
+	sensors       []string
+	// deprecated
 	deleteProcessedData bool
-	useLiveData         bool
+	// deprecated
+	useLiveData bool
 
+	modularizationV2Enabled bool
+	cartofacade             *cartofacade.CartoFacade
+	cartoFacadeTimeout      time.Duration
+
+	// deprecated
 	port       string
 	dataRateMs int
 	mapRateSec int
 
+	// deprecated
 	cancelFunc              func()
+	cancelSensorProcessFunc func()
+	cancelCartoFacadeFunc   func()
 	logger                  golog.Logger
+	// deprecated
 	activeBackgroundWorkers sync.WaitGroup
-
-	bufferSLAMProcessLogs        bool
-	slamProcessLogReader         io.ReadCloser
-	slamProcessLogWriter         io.WriteCloser
+	sensorProcessWorkers    sync.WaitGroup
+	cartoFacadeWorkers      sync.WaitGroup
```

Comment:
Keeping separate contexts for these allow us to safely shut down the sensor process, then stopping & terminating the carto facade object, before finally shutting down the carto facade background go routine.